Ruby/Ruby.sublime-syntax

```diff
@@ -330,22 +329,77 @@ contexts:
             - include: expressions
 
   keywords:
-    - match: '\b(BEGIN|END)\b(?![?!])'
-      scope: keyword.control.ruby
-    - match: '\b(class|module)\b(?![?!])'
-      scope: keyword.control.ruby
-    - match: '\b(begin|end|ensure|rescue)\b(?![?!])'
-      scope: keyword.control.ruby
-    - match: '\b(case|else|ensure|for|in|then)\b(?![?!])'
-      scope: keyword.control.ruby
-    - match: '\b(elsif|if|unless|when|while|until)\b(?![?!])'
-      scope: keyword.control.ruby
+    # blocks
+    - match: \bbegin\b(?![?!])
+      scope: keyword.control.block.begin.ruby
+      push: after-keyword
+    - match: \bend\b(?![?!])
+      scope: keyword.control.block.end.ruby
+    # conditional
+    - match: \bcase\b(?![?!])
+      scope: keyword.control.conditional.case.ruby
+      push: after-keyword
+    - match: \belse\b(?![?!])
+      scope: keyword.control.conditional.if.ruby
+      push: after-keyword
+    - match: \belsif\b(?![?!])
+      scope: keyword.control.conditional.elseif.ruby
+      push: after-keyword
+    - match: \bif\b(?![?!])
+      scope: keyword.control.conditional.if.ruby
+      push: after-keyword
+    - match: \bthen\b(?![?!])
+      scope: keyword.control.conditional.then.ruby
+    - match: \bunless\b(?![?!])
+      scope: keyword.control.conditional.unless.ruby
+      push: after-keyword
+    - match: \bwhen\b(?![?!])
+      scope: keyword.control.conditional.when.ruby
+      push: after-keyword
+    # exception
+    - match: \bensure\b(?![?!])
+      scope: keyword.control.exception.ensure.ruby
+      push: after-keyword
+    - match: \brescue\b(?![?!])
+      scope: keyword.control.exception.rescue.ruby
+      push: after-keyword
+    # loop
+    - match: \bfor\b(?![?!])
+      scope: keyword.control.loop.for.ruby
+      push: after-keyword
+    - match: \buntil\b(?![?!])
+      scope: keyword.control.loop.until.ruby
       push: after-keyword
-    - match: \b(and|not|or)\b
+    - match: \bwhile\b(?![?!])
+      scope: keyword.control.loop.while.ruby
+      push: after-keyword
+    # flow
+    - match: \bbreak\b(?![?!])
+      scope: keyword.control.flow.break.ruby
+    - match: \bnext\b(?![?!])
+      scope: keyword.control.flow.next.ruby
+    - match: \bredo\b(?![?!])
+      scope: keyword.control.flow.redo.ruby
+    - match: \bretry\b(?![?!])
+      scope: keyword.control.flow.retry.ruby
+    - match: \breturn\b(?![?!])
+      scope: keyword.control.flow.return.ruby
+    - match: \bsuper(?![?!])
+      scope: keyword.control.flow.super.ruby
```

Comment:
We haven't talked about `super` in #1228, but I guess this is an okay scope for how it appears to be used in Ruby (calls the overridden function with given parameters).

What do you think about C++ or Java?

Reply:
I am not sure about `super` actually. It sounds like a special keyword to tell ruby to call the ancestor method. In other languages it's sometimes more like a special variable compared to `self` or `this`. Haven't investigated all the different languages so far.

Reply:
Function calls *are* control flow, but I guess this most closely relates to `yield` (and maybe `async`). The only other scope I can think of right now would be something in `keyword.other`, which isn't ideal. No strong opinion.

Reply:
>  sounds like a special keyword to tell ruby to call the ancestor method

That is exactly what it is. Or rather, it's a special *function* which points to the ancestor method. I would actually *almost* consider scoping it as `support.function`. I don't think `keyword.control.flow` is appropriate. `keyword.other.super` is not terrible. @FichteFoll is right that it's similar to Python's `yield`.

Reply:
With keywords being used very often, I find `support.function` a good choice here. It also makes clear a function is called.

Reply:
... but the ruby docs name [super](https://ruby-doc.org/docs/keywords/1.9/files/keywords_rb.src/M000034.html) as keyword.

Reply:
A bit confusing is to compare `super` with `yield` which is a `keyword.control.flow` but then argue to not scope `super` as such.

Reply:
Keywords are special because they have a high priority in syntax parsing and can't be used as identifiers. Still, `super` and `this`, in some languages, just refer to different *things* depending on the context. 

- In Java and C#, `super` kind of behaves like `this` and you use it as `super.printMethod();` or just `super()` for the constructor. 
- In Ruby, it behaves like a magic function that resolves to the appropriate function. 
- In Python, it is an actual function that does these two things combined.

Afaik, we generally scope `this` as `variable.language` based on TM's guidelines, despite it being a keyword. I don't know if this is the best choice, but it's what we have currently, at least. If this is fine, then we could use `support.function.{language,super}` **if** it resembles a function call, like it does here. For Java and C# it would then be scoped `variable.language`. 

We could assign these scopes on top of `keyword.other` to represent the fact that they *are* keywords.

---

I don't think the comparison with `yield` was a good one. The only behavior they share is that they exit the current context at the position of the keyword and give a re-entry point at the same position. Ruby's `super` behaves more like a function call. Either way, it's not `keyword.control.flow`

Reply:
I think we can follow Python and use `support.function.builtin` here. As keywords are already used very often and with regards to `this` and `self`, I don't think we need to stack it with `keyword.other`.

Reply:
>As keywords are already used very often and with regards to this and self, I don't think we need to stack it with keyword.other.

Could you explain this in more detail? I don't see how keywords frequently being used in a language would be related to not using a `keyword` scope for a keyword.

Reply:
see commit message a11cf6d

Reply:
Well, let's tackle this at another time.

Reply:
Well, let's tackle this at another time.